src/typeset_settings_dialog.cpp

```diff
@@ -0,0 +1,96 @@
+#include "typeset_settings_dialog.h"
+#include "ui_typeset_settings_dialog.h"
+
+#include "forscape_common.h"
+#include <typeset_themes.h>
+
+#ifndef NDEBUG
+#include <iostream>
+#endif
+
+namespace Forscape {
+
+static std::array<QComboBox*, Code::NUM_SETTINGS> combo_boxes = { 0 };
+
+SettingsDialog* SettingsDialog::instance = nullptr;
+
+SettingsDialog::SettingsDialog(QWidget *parent) :
+    QDialog(parent),
+    ui(new Ui::SettingsDialog) {
+    ui->setupUi(this);
+
+    setWindowFlags(windowFlags() & ~Qt::WindowContextHelpButtonHint);
+
+    // DO THIS: codegen and have better names
+    combo_boxes[0] = ui->comboBox;
+    combo_boxes[1] = ui->comboBox_2;
+
+    for(QComboBox* combo_box : combo_boxes)
+        connect(combo_box, SIGNAL(currentIndexChanged(int)), this, SLOT(updateBackgroundColour()));
+}
+
+SettingsDialog::~SettingsDialog() {
+    delete ui;
+}
+
+static QColor background_colour[Code::NUM_WARNING_LEVELS+1];
+static QColor text_colour[Code::NUM_WARNING_LEVELS+1];
+
+bool SettingsDialog::execSettingsForm(const std::vector<Code::Settings::Update>& settings) noexcept {
+    if(instance == nullptr) instance = new SettingsDialog;
+
+    //Set colour coding
+    background_colour[0] = instance->palette().color(QPalette::ColorRole::Base);
+    background_colour[1+Code::NO_WARNING] = Typeset::getColour(Typeset::COLOUR_HIGHLIGHTSECONDARY);
+    background_colour[1+Code::WARN] = Typeset::getColour(Typeset::COLOUR_WARNINGBACKGROUND);
+    background_colour[1+Code::ERROR] = Typeset::getColour(Typeset::COLOUR_ERRORBACKGROUND);
+    text_colour[0] = instance->palette().color(QPalette::ColorRole::Text);
+    text_colour[1+Code::NO_WARNING] = Typeset::getColour(Typeset::COLOUR_TEXT);
+    text_colour[1+Code::WARN] = Typeset::getColour(Typeset::COLOUR_ID);
+    text_colour[1+Code::ERROR] = Typeset::getColour(Typeset::COLOUR_ID);
+    for(QComboBox* combo_box : combo_boxes){
+        for(int i = 0; i < Code::NUM_WARNING_LEVELS+1; i++){
+            combo_box->setItemData(i, background_colour[i], Qt::ItemDataRole::BackgroundRole);
+            combo_box->setItemData(i, text_colour[i], Qt::ItemDataRole::ForegroundRole);
+        }
+    }
+
+    //Set active index for each entry
+    for(QComboBox* combo_box : combo_boxes){
+        combo_box->setFocus();
+        combo_box->setCurrentIndex(0);
+    }
+    for(const Code::Settings::Update& update : settings){
+        combo_boxes[update.setting_id]->setFocus();
+        combo_boxes[update.setting_id]->setCurrentIndex( update.prev_value + 1 );
+    }
+
+    return instance->exec() == QDialog::Accepted;
```

Comment:
Not so fast there, bucko! You need to check for a null update. This would be a good time to assert that every Command changes the document, and undo/redo or redo/undo don't change the document

Reply:
Done on all counts